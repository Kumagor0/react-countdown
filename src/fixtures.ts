type TMockDateNow = {
  readonly now: jest.Mock;
  readonly timeDiff: number;
};

export function mockDateNow(): TMockDateNow {
  const timeDiff = 90110456;
  const now = jest.fn(() => 1482363367071);
  Date.now = now;

  return { now, timeDiff };
}

let mockTimerId: number;

export function mockRaf(): void {
  jest.spyOn(window, 'requestAnimationFrame').mockImplementation((fn) => {
    mockTimerId = window.setTimeout(fn, 1000);
    return mockTimerId;
  });

  jest
    .spyOn(window, 'cancelAnimationFrame')
    .mockImplementation(() => window.clearTimeout(mockTimerId));
}

export const defaultStats = {
  total: 0,
  days: 0,
  hours: 0,
  minutes: 0,
  seconds: 0,
  milliseconds: 0,
  completed: false,
};
