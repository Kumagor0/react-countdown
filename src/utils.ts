export type CountdownTimeUnit = 'd' | 'h' | 'm' | 's' | 'ms';

export interface CountdownTimeUnits<T = number> {
  readonly days: T;
  readonly hours: T;
  readonly minutes: T;
  readonly seconds: T;
  readonly milliseconds: T;
}

export interface CountdownTimeDeltaOptions {
  readonly now?: () => number;
  readonly unit?: CountdownTimeUnit;
  readonly precision?: number;
  readonly controlled?: boolean;
  readonly offsetTime?: number;
}

export interface CountdownTimeDelta extends CountdownTimeUnits {
  readonly total: number;
  readonly completed: boolean;
}

export interface CountdownTimeDeltaFormatted extends CountdownTimeUnits<string> {}

export interface CountdownTimeDeltaFormatOptions {
  readonly zeroPadDays?: number;
  readonly zeroPadTime?: number;
}

/**
 * Pads a given string or number with zeros.
 *
 * @export
 * @param {number|string} value Value to zero-pad.
 * @param {number} [length=2] Amount of characters to pad.
 * @returns Left-padded number/string.
 */
export function zeroPad(value: number | string, length: number = 2): string {
  const strValue = String(value);
  if (length === 0) return strValue;
  const match = strValue.match(/(.*?)([0-9]+)(.*)/);
  const prefix = match ? match[1] : '';
  const suffix = match ? match[3] : '';
  const strNo = match ? match[2] : strValue;
  const paddedNo =
    strNo.length >= length
      ? strNo
      : ([...Array(length)].map(() => '0').join('') + strNo).slice(length * -1);
  return `${prefix}${paddedNo}${suffix}`;
}

export const timeDeltaFormatOptionsDefaults: CountdownTimeDeltaFormatOptions = {
  zeroPadTime: 2,
};

/**
 * Calculates each time unit of a given timestamp.
 *
 * @export
 * @param {number} timestamp Timestamp of a certain point in time.
 * @returns {CountdownTimeUnits} Object that includes details about each time unit.
 */
export function calcTimeUnits(
  timestamp: number,
  unit: CountdownTimeUnit = 'h'
): CountdownTimeUnits {
  const total = timestamp / 1000;
  let days = Math.floor(total / (3600 * 24));
  let hours = Math.floor((total / 3600) % 24);
  let minutes = Math.floor((total / 60) % 60);
  let seconds = Math.floor(total % 60);
  let milliseconds = Number(((total % 1) * 1000).toFixed());
  console.log(unit);
  if (unit) {
    if (unit !== 'd') {
      hours += days * 24;
      days = 0;
    }

    if (unit !== 'd' && unit !== 'h') {
      minutes += hours * 60;
      hours = 0;
    }

    if (unit !== 'd' && unit !== 'h' && unit !== 'm') {
      seconds += minutes * 60;
      minutes = 0;
    }

    if (unit !== 'd' && unit !== 'h' && unit !== 'm' && unit !== 's') {
      milliseconds += seconds * 1000;
      seconds = 0;
    }
  }

  return {
    days,
    hours,
    minutes,
    seconds,
    milliseconds,
  };
}

/**
 * Formats a given countdown time units object.
 *
 * @export
 * @param {CountdownTimeUnits} units
 * @param {CountdownTimeDeltaFormatOptions} [options]
 * @returns {CountdownTimeDeltaFormatted} Formatted time units object.
 */
export function formatTimeUnits(
  delta: CountdownTimeUnits,
  options?: CountdownTimeDeltaFormatOptions
): CountdownTimeDeltaFormatted {
  const { days, hours, minutes, seconds, milliseconds } = delta;
  const { zeroPadTime, zeroPadDays = zeroPadTime } = {
    ...timeDeltaFormatOptionsDefaults,
    ...options,
  };

  const zeroPadLength = Math.min(2, zeroPadTime);

  return {
    days: zeroPad(days, zeroPadDays),
    hours: zeroPad(hours, zeroPadLength),
    minutes: zeroPad(minutes, zeroPadLength),
    seconds: zeroPad(seconds, zeroPadLength),
    milliseconds: zeroPad(milliseconds, zeroPadLength),
  };
}

/**
 * Calculates the time difference between a given end date and the current date.
 *
 * @export
 * @param {Date|number|string} date Date or timestamp representation of the end date.
 * @param {Object} [{ now = Date.now, precision = 0, controlled = false }={}]
 *  {function} [date=Date.now] Alternative function for returning the current date.
 *  {number} [precision=0] The precision on a millisecond basis.
 *  {boolean} [controlled=false] Defines whether the calculated value is already provided as the time difference or not.
 *  {number} [offsetTime=0] Defines the offset time that gets added to the start time; only considered if controlled is false.
 * @param {number} [precision=0] The precision on a millisecond basis.
 * @param {boolean} [controlled=false] Defines whether the calculated value is already provided as the time difference or not.
 * @returns {CountdownTimeDelta} Object that includes details about the time difference.
 */
export function calcTimeDelta(
  date: Date | string | number,
  {
    now = Date.now,
    unit = 'h',
    precision = 0,
    controlled = false,
    offsetTime = 0,
  }: CountdownTimeDeltaOptions = {}
): CountdownTimeDelta {
  let startTimestamp: number;

  if (date instanceof Date) {
    startTimestamp = date.getTime();
  } else if (typeof date === 'string') {
    startTimestamp = new Date(date).getTime();
  } else {
    startTimestamp = date;
  }

  if (!controlled) {
    startTimestamp += offsetTime;
  }

  const total = Math.round(
    parseFloat(
      (Math.max(0, controlled ? startTimestamp : startTimestamp - now()) / 1000).toFixed(
        Math.max(0, Math.min(20, precision))
      )
    ) * 1000
  );

  return {
    ...calcTimeUnits(total, unit),
    total,
    completed: total <= 0,
  };
}
