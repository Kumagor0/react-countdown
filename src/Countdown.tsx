import React from 'react';
import PropTypes from 'prop-types';

import CountdownJs, { CountdownProps, CountdownState, CountdownApi } from './CountdownJs';

/**
 * A customizable countdown component for React.
 *
 * @export
 * @class Countdown
 * @extends {React.Component}
 */
export default class Countdown extends React.Component<CountdownProps, CountdownState> {
  countdown: CountdownJs;

  static propTypes = {
    date: PropTypes.oneOfType([PropTypes.instanceOf(Date), PropTypes.string, PropTypes.number])
      .isRequired,
    unit: PropTypes.oneOf(['d', 'h', 'm', 's', 'ms']),
    zeroPadTime: PropTypes.number,
    zeroPadDays: PropTypes.number,
    controlled: PropTypes.bool,
    raf: PropTypes.bool,
    intervalDelay: PropTypes.number,
    precision: PropTypes.number,
    autoStart: PropTypes.bool,
    children: PropTypes.element,
    renderer: PropTypes.func,
    now: PropTypes.func,
    pure: PropTypes.bool,
    onMount: PropTypes.func,
    onStart: PropTypes.func,
    onPause: PropTypes.func,
    onTick: PropTypes.func,
    onComplete: PropTypes.func,
  };

  constructor(props: CountdownProps) {
    super(props);
    this.countdown = new CountdownJs(this.props, (state, callback) =>
      this.setState(state, callback)
    );
  }

  componentDidMount(): void {
    this.countdown.init();
  }

  componentDidUpdate(): void {
    this.countdown.update(this.props);
  }

  componentWillUnmount(): void {
    this.countdown.destroy();
  }

  getApi = (): CountdownApi => {
    return this.countdown.getApi();
  };

  renderUnit(unit: CountdownProps['unit']): React.ReactNode {
    return null;
  }

  render(): React.ReactNode {
    const { children, renderer } = this.props;
    const renderProps = this.countdown.getRenderProps();

    if (renderer) {
      return renderer(renderProps);
    }

    if (children && renderProps.completed) {
      return React.cloneElement(children, { countdown: renderProps });
    }

    const { days, hours, minutes, seconds } = renderProps.formatted;
    return (
      <>
        {days}
        {days ? ':' : ''}
        {hours}:{minutes}:{seconds}
      </>
    );
  }
}
