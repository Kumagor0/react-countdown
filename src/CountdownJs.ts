import Countdown from './Countdown';
import {
  calcTimeDelta,
  CountdownTimeDelta,
  CountdownTimeDeltaFormatted,
  CountdownTimeDeltaFormatOptions,
  timeDeltaFormatOptionsDefaults,
  formatTimeUnits,
} from './utils';

export interface CountdownProps extends React.Props<Countdown>, CountdownTimeDeltaFormatOptions {
  readonly date: Date | number | string;
  readonly key?: React.Key;
  readonly controlled?: boolean;
  readonly raf?: boolean;
  readonly intervalDelay?: number;
  readonly unit?: 'd' | 'h' | 'm' | 's' | 'ms';
  readonly precision?: number;
  readonly autoStart?: boolean;
  readonly children?: React.ReactElement<unknown>;
  readonly renderer?: (props: CountdownRenderProps) => React.ReactNode;
  readonly now?: () => number;
  readonly pure?: boolean;
  readonly onMount?: CountdownTimeDeltaFn;
  readonly onStart?: CountdownTimeDeltaFn;
  readonly onPause?: CountdownTimeDeltaFn;
  readonly onTick?: CountdownTimeDeltaFn;
  readonly onComplete?: CountdownTimeDeltaFn;
}

export interface CountdownRenderProps extends CountdownTimeDelta {
  readonly api: CountdownApi;
  readonly props: CountdownProps;
  readonly formatted: CountdownTimeDeltaFormatted;
}

export type CountdownTimeDeltaFn = (timeDelta: CountdownTimeDelta) => void;

export interface CountdownState {
  readonly timeDelta: CountdownTimeDelta;
  readonly offsetStart: number;
  readonly offsetTime: number;
}

export interface CountdownApi {
  readonly start: () => void;
  readonly pause: () => void;
  readonly isPaused: () => boolean;
  readonly isCompleted: () => boolean;
}

export type StateUpdaterFn = (state: CountdownState, callback?: () => void) => void;

/**
 * A customizable countdown component for React.
 *
 * @export
 * @class Countdown
 * @extends {React.Component}
 */
export default class CountdownJs {
  props: CountdownProps;
  state: CountdownState;
  stateUpdater: StateUpdaterFn;

  initialized = false;
  timer: number | undefined;
  api: CountdownApi | undefined;

  constructor(props: CountdownProps, stateUpdater: StateUpdaterFn) {
    this.props = this.computeProps(props);
    this.state = {
      timeDelta: this.calcTimeDelta(),
      offsetStart: this.props.autoStart ? 0 : this.calcOffsetStart(),
      offsetTime: 0,
    };
    this.stateUpdater = stateUpdater;
  }

  init = (): void => {
    this.initialized = true;
    this.props.autoStart && this.start();
    this.props.onMount && this.props.onMount(this.calcTimeDelta());
  };

  update = (props: CountdownProps): boolean => {
    const nextProps = this.computeProps(props);

    if (nextProps.pure && !this.shallowCompareProps(nextProps, this.computeProps(this.props))) {
      this.setProps(nextProps);
      this.setTimeDeltaState(this.calcTimeDelta());
      return true;
    }

    return false;
  };

  destroy = (): void => {
    this.initialized = false;
    this.clearTimer();
  };

  tick = (): void => {
    const { onTick } = this.props;
    const timeDelta = this.calcTimeDelta();

    this.setTimeDeltaState(timeDelta);

    if (onTick && timeDelta.total > 0) {
      onTick(timeDelta);
    }
  };

  rafTick = (): void => {
    this.tick();

    if (this.timer) {
      this.timer = requestAnimationFrame(this.rafTick);
    }
  };

  calcTimeDelta(): CountdownTimeDelta {
    const { date, now, unit, precision, controlled } = this.props;
    return calcTimeDelta(date, {
      now,
      unit,
      precision,
      controlled,
      offsetTime: this.state ? this.state.offsetTime : 0,
    });
  }

  calcOffsetStart(): number {
    return Date.now();
  }

  start = (): void => {
    this.setState(
      ({ offsetStart, offsetTime }: CountdownState) => ({
        offsetStart: 0,
        offsetTime: offsetTime + (offsetStart ? this.calcOffsetStart() - offsetStart : 0),
      }),
      () => {
        const timeDelta = this.calcTimeDelta();
        this.setTimeDeltaState(timeDelta);
        this.props.onStart && this.props.onStart(timeDelta);

        if (!this.props.controlled) {
          this.timer = this.startTimer();
        }
      }
    );
  };

  pause = (): void => {
    this.clearTimer();
    this.setState({ offsetStart: this.calcOffsetStart() }, () => {
      const timeDelta = this.calcTimeDelta();
      this.setTimeDeltaState(timeDelta);
      this.props.onPause && this.props.onPause(timeDelta);
    });
  };

  startTimer(): number {
    this.clearTimer();
    return this.props.raf
      ? window.requestAnimationFrame(this.rafTick)
      : window.setInterval(this.tick, this.props.intervalDelay);
  }

  clearTimer(): void {
    if (this.timer) {
      if (this.props.raf) {
        window.cancelAnimationFrame(this.timer);
      } else {
        window.clearInterval(this.timer);
      }

      this.timer = undefined;
    }
  }

  computeProps(props: CountdownProps): CountdownProps {
    return {
      ...timeDeltaFormatOptionsDefaults,
      controlled: false,
      raf: true,
      intervalDelay: 1000,
      precision: 0,
      autoStart: true,
      pure: true,
      ...props,
    };
  }

  isPaused = (): boolean => {
    return this.state.offsetStart > 0;
  };

  isCompleted = (): boolean => {
    return this.state.timeDelta.completed;
  };

  shallowCompareProps(propsA: CountdownProps, propsB: CountdownProps): boolean {
    const keysA = Object.keys(propsA);
    return (
      keysA.length === Object.keys(propsB).length &&
      !keysA.some(keyA => {
        const valueA = propsA[keyA];
        const valueB = propsB[keyA];
        return (
          !propsB.hasOwnProperty(keyA) ||
          !(valueA === valueB || (valueA !== valueA && valueB !== valueB)) // NaN !== NaN
        );
      })
    );
  }

  setTimeDeltaState(timeDelta: CountdownTimeDelta): void {
    let callback;

    if (!this.isCompleted() && timeDelta.completed) {
      this.clearTimer();

      callback = (): void => this.props.onComplete && this.props.onComplete(timeDelta);
    }

    if (this.initialized) {
      this.setState({ timeDelta }, callback);
    }
  }

  getApi = (): CountdownApi => {
    return (this.api = this.api || {
      start: this.start,
      pause: this.pause,
      isPaused: this.isPaused,
      isCompleted: this.isCompleted,
    });
  };

  getRenderProps = (): CountdownRenderProps => {
    const { zeroPadTime, zeroPadDays } = this.props;
    const { timeDelta } = this.state;
    return {
      ...timeDelta,
      api: this.getApi(),
      props: this.props,
      formatted: formatTimeUnits(timeDelta, {
        zeroPadTime,
        zeroPadDays,
      }),
    };
  };

  getProps = (): CountdownProps => {
    return this.props;
  };

  setProps(props: CountdownProps): void {
    this.props = this.computeProps(props);
  }

  getState = (): CountdownState => {
    return this.state;
  };

  setState = (
    state: Partial<CountdownState> | ((prevState: CountdownState) => Partial<CountdownState>),
    callback?: () => void
  ): void => {
    this.state = {
      ...this.state,
      ...(typeof state === 'function' ? state(this.state) : state),
    };

    this.stateUpdater(this.state, callback);
  };
}
