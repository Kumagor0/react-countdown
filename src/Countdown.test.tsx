import * as React from 'react';
import { mount, ReactWrapper } from 'enzyme';

import Countdown from './Countdown';
import CountdownJs, { CountdownProps, CountdownState } from './CountdownJs';
import { calcTimeDelta, formatTimeUnits, CountdownTimeUnit } from './utils';
import { mockDateNow, defaultStats } from './fixtures';

const { now, timeDiff } = mockDateNow();

describe('<Countdown />', () => {
  jest.useFakeTimers();

  let wrapper: ReactWrapper<CountdownProps, CountdownState, Countdown>;
  let obj: Countdown;
  let countdownDate: number;
  const countdownMs = 10000;

  beforeEach(() => {
    Date.now = now;
    const date = Date.now() + countdownMs;
    countdownDate = date;
    wrapper = mount(<Countdown date={date} />);
    obj = wrapper.instance();
  });

  function getCountdownJsInstance(): CountdownJs {
    return wrapper.instance().countdown;
  }

  function getCountdownJsState(): CountdownState {
    return getCountdownJsInstance().getState();
  }

  it('should render a simple countdown', () => {
    wrapper = mount(<Countdown date={Date.now() + timeDiff} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render custom renderer output', () => {
    wrapper = mount(
      <Countdown
        date={Date.now() + timeDiff}
        renderer={({ days, hours, minutes, seconds, milliseconds }) => (
          <div>{`${days}d:${hours}h:${minutes}m:${seconds}s:${milliseconds}ms`}</div>
        )}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render and unmount component on countdown end', () => {
    const zeroPadTime = 0;

    class Completionist extends React.PureComponent<{
      readonly name: string;
      readonly children: React.ReactNode;
    }> {
      render() {
        const { name, children } = this.props;
        return (
          <div>
            Completed! {name} {children}
          </div>
        );
      }
    }

    let completionist;
    Completionist.prototype.componentDidMount = jest.fn();

    wrapper = mount(
      <Countdown date={Date.now() + timeDiff} zeroPadTime={zeroPadTime}>
        <Completionist
          ref={el => {
            completionist = el;
          }}
          name="master"
        >
          Another child
        </Completionist>
      </Countdown>
    );
    expect(Completionist.prototype.componentDidMount).not.toBeCalled();
    expect(wrapper).toMatchSnapshot();

    // Forward in time
    wrapper.setProps({ date: 0 });
    expect(getCountdownJsState().timeDelta.completed).toBe(true);
    expect(wrapper.props().children!.type).toBe(Completionist);
    expect(Completionist.prototype.componentDidMount).toBeCalled();

    const computedProps = { ...wrapper.props() };
    delete computedProps.children;

    obj = wrapper.instance();
    const delta = getCountdownJsState().timeDelta;
    expect(completionist.props).toEqual({
      countdown: {
        ...delta,
        api: obj.getApi(),
        props: getCountdownJsInstance().getProps(),
        formatted: formatTimeUnits(delta, { zeroPadTime }),
      },
      name: 'master',
      children: 'Another child',
    });
    expect(wrapper).toMatchSnapshot();
  });

  describe('unit', () => {
    (['d', 'h', 'm', 's', 'ms'] as CountdownTimeUnit[]).forEach(unit =>
      it.only(`should render with unit => "${unit}"`, () => {
        wrapper = mount(
          <Countdown
            date={Date.now() + timeDiff}
            unit={unit}
            renderer={({ days, hours, minutes, seconds, milliseconds }) => (
              <div>{`${days}d:${hours}h:${minutes}m:${seconds}s:${milliseconds}ms`}</div>
            )}
          />
        );
        expect(wrapper).toMatchSnapshot();
      })
    );
  });

  it('should render with zeroPadDays => 3', () => {
    wrapper = mount(<Countdown date={Date.now() + 10 * 86400 * 1000} zeroPadDays={3} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should trigger onTick and onComplete callbacks', () => {
    const onTick = jest.fn(stats => {
      expect(stats).toEqual(calcTimeDelta(countdownDate));
    });

    const onComplete = jest.fn(stats => {
      expect(stats.total).toEqual(0);
    });

    wrapper.setProps({ onTick, onComplete });
    expect(onTick).not.toBeCalled();

    // Forward 6s in time
    now.mockReturnValue(countdownDate - 6000);
    jest.runTimersToTime(6000);
    expect(onTick.mock.calls.length).toBe(6);
    expect(getCountdownJsState().timeDelta.total).toBe(6000);

    wrapper.update();
    expect(wrapper).toMatchSnapshot();

    // Forward 3 more seconds
    now.mockReturnValue(countdownDate - 1000);
    jest.runTimersToTime(3000);
    expect(onTick.mock.calls.length).toBe(9);
    expect(getCountdownJsState().timeDelta.total).toBe(1000);
    expect(getCountdownJsState().timeDelta.completed).toBe(false);

    // The End: onComplete callback gets triggered instead of onTick
    now.mockReturnValue(countdownDate);
    jest.runTimersToTime(1000);
    expect(onTick.mock.calls.length).toBe(9);
    expect(onTick).toBeCalledWith({
      ...defaultStats,
      total: 1000,
      seconds: 1,
    });

    expect(onComplete.mock.calls.length).toBe(1);
    expect(onComplete).toBeCalledWith({ ...defaultStats, completed: true });
    expect(getCountdownJsState().timeDelta.completed).toBe(true);
  });

  it('should run through the controlled component by updating the date prop', () => {
    wrapper = mount(<Countdown date={1000} controlled />);
    obj = wrapper.instance();
    const api = obj.getApi();
    const countdownJsObj = getCountdownJsInstance();

    expect(countdownJsObj.timer).toBeUndefined();
    expect(getCountdownJsState().timeDelta.completed).toBe(false);
    expect(api.isCompleted()).toBe(false);

    wrapper.setProps({ date: 0 });
    expect(getCountdownJsState().timeDelta.total).toBe(0);
    expect(getCountdownJsState().timeDelta.completed).toBe(true);
    expect(api.isCompleted()).toBe(true);
  });

  it('should only re-set time delta state when props have changed', () => {
    const root = document.createElement('div');
    wrapper = mount(<Countdown date={1000} />, { attachTo: root });
    const countdownJsObj = getCountdownJsInstance();
    countdownJsObj.setTimeDeltaState = jest.fn();

    function mergeProps(partialProps: Partial<CountdownProps>): CountdownProps {
      return { ...wrapper.props(), ...partialProps };
    }

    wrapper.setProps(mergeProps({ date: 500 }));
    expect(countdownJsObj.setTimeDeltaState).toHaveBeenCalledTimes(1);

    wrapper.setProps(mergeProps({ intervalDelay: 999 }));
    expect(countdownJsObj.setTimeDeltaState).toHaveBeenCalledTimes(2);

    wrapper.setProps(mergeProps({ date: 500 }));
    expect(countdownJsObj.setTimeDeltaState).toHaveBeenCalledTimes(2);

    wrapper.setProps(mergeProps({ precision: NaN }));
    expect(countdownJsObj.setTimeDeltaState).toHaveBeenCalledTimes(3);

    wrapper.setProps(mergeProps({ precision: NaN }));
    expect(countdownJsObj.setTimeDeltaState).toHaveBeenCalledTimes(3);

    wrapper.setProps(mergeProps({ precision: 3 }));
    expect(countdownJsObj.setTimeDeltaState).toHaveBeenCalledTimes(4);

    wrapper.setProps(mergeProps({ date: 750 }));
    expect(countdownJsObj.setTimeDeltaState).toHaveBeenCalledTimes(5);
  });

  it('should not (try to) set state after component unmount', () => {
    const countdownJsObj = getCountdownJsInstance();
    expect(getCountdownJsState().timeDelta.completed).toBe(false);

    now.mockReturnValue(countdownDate - 6000);
    jest.runTimersToTime(6000);
    expect(countdownJsObj.initialized).toBe(true);
    expect(getCountdownJsState().timeDelta.total).toBe(6000);

    wrapper.unmount();

    now.mockReturnValue(countdownDate - 3000);
    jest.runTimersToTime(3000);
    expect(countdownJsObj.initialized).toBe(false);
    expect(countdownJsObj.getState().timeDelta.total).toBe(6000);

    countdownJsObj.setTimeDeltaState(defaultStats);
    expect(countdownJsObj.getState().timeDelta).not.toEqual(defaultStats);
  });

  it('should pause and restart countdown', () => {
    const spies = {
      onMount: jest.fn(),
      onStart: jest.fn(),
      onPause: jest.fn(),
    };
    wrapper = mount(<Countdown date={countdownDate} {...spies} />);
    obj = wrapper.instance();
    const api = obj.getApi();

    expect(getCountdownJsState()).toEqual(
      expect.objectContaining({
        offsetStart: 0,
        offsetTime: 0,
      })
    );
    expect(api.isPaused()).toBe(false);
    expect(spies.onMount).toHaveBeenCalledTimes(1);
    expect(spies.onMount).toHaveBeenCalledWith(getCountdownJsInstance().calcTimeDelta());
    expect(spies.onStart).toHaveBeenCalledTimes(1);
    expect(spies.onPause).toHaveBeenCalledTimes(0);

    let runMs = 2000;
    const nowBeforePause = countdownDate - (countdownMs - runMs);
    now.mockReturnValue(nowBeforePause);
    jest.runTimersToTime(runMs);
    expect(getCountdownJsState().timeDelta.total).toBe(countdownMs - runMs);

    api.pause();
    expect(api.isPaused()).toBe(true);
    expect(api.isCompleted()).toBe(false);
    expect(spies.onPause).toHaveBeenCalledTimes(1);
    expect(spies.onPause).toHaveBeenCalledWith(getCountdownJsInstance().calcTimeDelta());

    runMs += 2000;
    const pausedMs = 2000;
    now.mockReturnValue(countdownDate - (countdownMs - runMs));
    jest.runTimersToTime(runMs);
    expect(countdownMs - runMs + pausedMs).toBe(8000);
    expect(getCountdownJsState().timeDelta.total).toBe(8000);
    expect(getCountdownJsState()).toEqual(
      expect.objectContaining({
        offsetStart: nowBeforePause,
        offsetTime: 0,
      })
    );

    api.start();
    expect(api.isPaused()).toBe(false);
    expect(api.isCompleted()).toBe(false);
    expect(spies.onStart).toHaveBeenCalledTimes(2);
    expect(spies.onStart).toHaveBeenCalledWith(getCountdownJsInstance().calcTimeDelta());

    expect(getCountdownJsState().timeDelta.total).toBe(8000);
    expect(getCountdownJsState()).toEqual(
      expect.objectContaining({ offsetStart: 0, offsetTime: pausedMs })
    );

    runMs += 4000;
    now.mockReturnValue(countdownDate - (countdownMs - runMs));
    jest.runTimersToTime(runMs);
    expect(countdownMs - runMs + pausedMs).toBe(4000);
    expect(getCountdownJsState().timeDelta.total).toBe(4000);
    expect(getCountdownJsState()).toEqual(
      expect.objectContaining({
        offsetStart: 0,
        offsetTime: pausedMs,
      })
    );

    now.mockReturnValue(countdownDate + pausedMs);
    jest.runTimersToTime(countdownMs + pausedMs);
    expect(getCountdownJsState().timeDelta.completed).toBe(true);
    expect(api.isCompleted()).toBe(true);

    expect(spies.onMount).toHaveBeenCalledTimes(1);
    expect(spies.onPause).toHaveBeenCalledTimes(1);
    expect(spies.onStart).toHaveBeenCalledTimes(2);
  });

  it('should update component when pure', () => {
    wrapper = mount(<Countdown pure date={countdownDate} />);
    const countdownJsObj = getCountdownJsInstance();
    expect(countdownJsObj.getProps().date).toBe(countdownDate);

    wrapper.setProps({ date: 0 });
    expect(countdownJsObj.getProps().date).toBe(0);
  });

  it('should not update component when impure', () => {
    wrapper = mount(<Countdown pure={false} date={countdownDate} />);
    const countdownJsObj = getCountdownJsInstance();
    expect(countdownJsObj.getProps().date).toBe(countdownDate);

    wrapper.setProps({ date: 0 });
    expect(countdownJsObj.getProps().date).not.toBe(0);
  });

  it('should auto start countdown', () => {
    const spies = {
      onStart: jest.fn(),
      onPause: jest.fn(),
    };
    wrapper = mount(<Countdown date={countdownDate} autoStart={true} {...spies} />);
    const countdownObj = getCountdownJsInstance();
    const api = countdownObj.getApi();

    expect(spies.onStart).toHaveBeenCalledTimes(1);
    expect(spies.onPause).toHaveBeenCalledTimes(0);
    expect(api.isPaused()).toBe(false);
    expect(getCountdownJsInstance().getState()).toEqual(
      expect.objectContaining({
        offsetStart: 0,
        offsetTime: 0,
      })
    );

    api.pause();
    expect(spies.onStart).toHaveBeenCalledTimes(1);
    expect(spies.onPause).toHaveBeenCalledTimes(1);
    expect(api.isPaused()).toBe(true);
    expect(getCountdownJsInstance().getState()).toEqual(
      expect.objectContaining({
        offsetStart: countdownDate - countdownMs,
        offsetTime: 0,
      })
    );
  });

  it('should not auto start countdown', () => {
    const spies = {
      onStart: jest.fn(),
    };
    wrapper = mount(<Countdown date={countdownDate} autoStart={false} {...spies} />);
    const countdownObj = getCountdownJsInstance();
    const api = countdownObj.getApi();

    expect(spies.onStart).toHaveBeenCalledTimes(0);
    expect(api.isPaused()).toBe(true);
    expect(getCountdownJsInstance().getState()).toEqual(
      expect.objectContaining({
        offsetStart: countdownDate - countdownMs,
        offsetTime: 0,
      })
    );

    api.start();
    expect(spies.onStart).toHaveBeenCalledTimes(1);
    expect(api.isPaused()).toBe(false);
    expect(getCountdownJsInstance().getState()).toEqual(
      expect.objectContaining({
        offsetStart: 0,
        offsetTime: 0,
      })
    );
  });

  afterEach(() => {
    try {
      wrapper.detach();
    } catch (e) {
      // Continue...
    }
  });
});
