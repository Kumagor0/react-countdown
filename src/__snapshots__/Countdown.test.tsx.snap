// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Countdown /> should render a simple countdown 1`] = `
<Countdown
  date={1482453477527}
>
  01
  :
  01
  :
  01
  :
  50
</Countdown>
`;

exports[`<Countdown /> should render custom renderer output 1`] = `
<Countdown
  date={1482453477527}
  renderer={[Function]}
>
  <div>
    1
    1
    1
    50
  </div>
</Countdown>
`;

exports[`<Countdown /> should render with daysInHours => true 1`] = `
<Countdown
  date={1482453477527}
  daysInHours={true}
>
  25
  :
  01
  :
  50
</Countdown>
`;

exports[`<Countdown /> should render with zeroPadDays => 3 1`] = `
<Countdown
  date={1483227367071}
  zeroPadDays={3}
>
  010
  :
  00
  :
  00
  :
  00
</Countdown>
`;

exports[`<Countdown /> should trigger onTick and onComplete callbacks 1`] = `
<Countdown
  date={1482363377071}
  onComplete={[MockFunction]}
  onTick={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "completed": false,
            "days": 0,
            "hours": 0,
            "milliseconds": 0,
            "minutes": 0,
            "seconds": 6,
            "total": 6000,
          },
        ],
        Array [
          Object {
            "completed": false,
            "days": 0,
            "hours": 0,
            "milliseconds": 0,
            "minutes": 0,
            "seconds": 6,
            "total": 6000,
          },
        ],
        Array [
          Object {
            "completed": false,
            "days": 0,
            "hours": 0,
            "milliseconds": 0,
            "minutes": 0,
            "seconds": 6,
            "total": 6000,
          },
        ],
        Array [
          Object {
            "completed": false,
            "days": 0,
            "hours": 0,
            "milliseconds": 0,
            "minutes": 0,
            "seconds": 6,
            "total": 6000,
          },
        ],
        Array [
          Object {
            "completed": false,
            "days": 0,
            "hours": 0,
            "milliseconds": 0,
            "minutes": 0,
            "seconds": 6,
            "total": 6000,
          },
        ],
        Array [
          Object {
            "completed": false,
            "days": 0,
            "hours": 0,
            "milliseconds": 0,
            "minutes": 0,
            "seconds": 6,
            "total": 6000,
          },
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
>
  00
  :
  00
  :
  00
  :
  06
</Countdown>
`;
